extensions/amp-render/1.0/amp-render.js

```diff
@@ -158,6 +158,14 @@ export class AmpRender extends BaseElement {
 
   /** @override */
   init() {
+    this.registerApiAction('refresh', (api) => {
+      const src = this.element.getAttribute('src');
+      if (!src || isAmpStateSrc(src) || isAmpScriptSrc(src)) {
+        return;
```

Comment:
Suggestion: WDYT about storing `canRefresh` as an internal variable and updating in a `checkPropsPostMutations`? It doesn't seem necessary to calculate this every call unless `src` actually changes between refresh actions, but then again, maybe it is fine to repeat this logic. We also follow the same conditional branches in `getJsonFn`, would be nice if there was an even simpler way than what I suggested. 🤔

Reply:
`mutationObserverCallback` may also be helpful here if you plan to change the logic.  This method is called each time `attributes` change.  You can keep an instance variable of `src` and `canRefresh` (or however you'd like to store state) and it'll be synced with the `src` attribute on the element.

Here is an example usage: https://github.com/ampproject/amphtml/blob/main/extensions/amp-social-share/1.0/amp-social-share.js#L157

Depending on your config, the callback is also called on changes to `childList` and potentially other options so you might want to double check you update state only when needed.

Reply:
So mimicking the behavior of `amp-list` here - we should allow refreshing even though `src` is unchanged (there is an amp-list example in the docs on amp.dev).

Reply:
I thought about this a bit more - isn't is better to bail here on every `refresh` action rather than re-compute `canRefresh` on every `mutationObserverCallback` which involves iterating thru all the mutation records?

Reply:
You don't have to iterate through all mutation records, only read the final `getAttribute('src')` state:
```js
  mutationObserverCallback() {
    const open = this.element.getAttribute('src');
    if (src === this.src_) {
      return;
    }
    this.src_ = src;
    this.canRefresh_ = !src?.length || (...etc);
  }
```

Anyway, it is up to you and likely depends on how the author plans on using the component, i.e. if they never implement amp-render with `refresh` action but mutate `src` often, then it is more costly to compute in `mutationObserverCallback` and vice versa for placing the logic in the `registerApiAction` callback. The rest of the PR looks good to me so no need to block on this detail.

Reply:
I think either way is fine.  Just wanted to provide you another option.

I suppose it really comes down to which one is expected to change more or if you plan to use instance variable of `canRefresh` and `src` in other parts of the component.

Reply:
Appreciate the alternate solution here. I've added a comment for my future self linking to this thread, so I can revisit if necessary.